### PR TITLE
Enhancements for Carrier Billing v0.2 - Part I

### LINKED ISSUE #97: Improvements for Carrier Billing
**Problem description**

Several topics to be addressed. To be reviewed on next meeting.
Target version: Include within v0.2 or v0.3 (based on we agree to include within v0.2 or next one)

**- Enhancement in Look&Feel of Specific Exceptions:**

Example, instead of defining operation specific exceptions within exception code. Just define an object for them

`    PaymentPermissionDenied403:
      description: |
        Client does not have sufficient permission.
        In addition to regular PERMISSION_DENIED scenario other scenarios may exist:
          - Phone Number is required ("code": "CARRIER_BILLING.PHONE_NUMBER_REQUIRED","message": "Phone Number is required").
          - Phone Number provided not matching Authorization context ("code": "CARRIER_BILLING.INVALID_TOKEN_CONTEXT","message": "Phone Number does not match with Authorization context").
          - Unauthorized amount requested ("code": "CARRIER_BILLING.UNAUTHORIZED_AMOUNT","message": "Unauthorized amount requested").
          - Accumulated threshold amount for the user's mobile account overpassed ("code": "CARRIER_BILLING.USER_AMOUNT_THRESHOLD_OVERPASSED","message": "Unathorized payment request. Accumulated user mobile payments overpass account amount threshold").
      headers:
        x-correlator:
          $ref: '#/components/headers/x-correlator'      
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/ErrorInfo"
          examples:
            Generic403:
              summary: Forbidden
              value:
                status: 403
                code: PERMISSION_DENIED
                message: "Operation not allowed: ..."
            PhoneNumberRequired:
              summary: Phone Number not provided and cannot be obtained from Authorization context
              value:
                status: 403
                code: CARRIER_BILLING.PHONE_NUMBER_REQUIRED
                message: "Phone Number is required"
            PhoneNumberMismatch:
              summary: Phone Number provided not matching Authorization context
              value:
                status: 403
                code: CARRIER_BILLING.INVALID_TOKEN_CONTEXT
                message: "Phone Number does not match with Authorization context"
            UnauthorizedAmount:
              summary: Unauthorized amount requested
              value:
                status: 403
                code: CARRIER_BILLING.UNAUTHORIZED_AMOUNT
                message: "Unauthorized amount requested"
            UserMobileAccumulatedThresholdAmountOverpassed:
              summary: Accumulated threshold amount for the user's mobile account overpassed
              value:
                status: 403
                code: CARRIER_BILLING.USER_AMOUNT_THRESHOLD_OVERPASSED
                message: "Unathorized payment request. Accumulated user mobile payments over` 

Then look and feel is better for API devs:

image

**- Notifications endpoint alignment:**

* As Commonalities issue #8 opened: https://github.com/camaraproject/Commonalities/issues/8.
It makes sense to remove specific Notifications endpoint and only refer to callback in POST actions. 
In the same fashion like QoD: https://github.com/camaraproject/QualityOnDemand/blob/main/code/API_definitions/qod-api.yaml#L99

### PR DESCRIPTION
**Pull request by PedroDiez**
#### What type of PR is this?

Add one of the following kinds:

* enhancement/feature
* documentation

#### What this PR does / why we need it:

This PR deals with some Issues whose output is already agreed:
- [Issue#97](https://github.com/camaraproject/CarrierBillingCheckOut/issues/97) [COVERED]
- [Issue#105](https://github.com/camaraproject/CarrierBillingCheckOut/issues/105) [COVERED: Impacts in amount and taxAmount]
- [Issue#106](https://github.com/camaraproject/CarrierBillingCheckOut/issues/106) [COVERED: Inspired in `/documentation/API_documentation/Carrier_Billing_API.md` and aligned with current model approach for QoD and Home Devices QoD APIs. Providing details about state transitions -in the current context, without considerations of Issue#116]
- [Issue#107](https://github.com/camaraproject/CarrierBillingCheckOut/issues/107) [COVERED]
- [Issue#108](https://github.com/camaraproject/CarrierBillingCheckOut/issues/108) [COVERED]
- [Issue#112](https://github.com/camaraproject/CarrierBillingCheckOut/issues/112) [COVERED]

#### Which issue(s) this PR fixes:

Fixes #97, #105, #106, #107, #108, #112

#### Special notes for reviewers:

**COMPLETED**. Ready for review

### CODE REVIEW

**Comment by bigludo7** [comment]
Thanks @PedroDiez 
I saw the fixes for 107 & 108 - Looks good for me. Do you want me to approve or wait for the other issue inside this PR to be solved?
Thanks

**Comment by PedroDiez** [comment]
Hi @bigludo7, Many thanks for the review!. I wanted to add these ones in advance as were easy to manage.
I prefer waiting for final review until i include the rest of points.

**Comment by PedroDiez** [comment]
Ready for review: @bigludo7, @rartych, @alabajnaid

**Review by rartych — COMMENTED** [review]
Current [API Design Guidelines](https://github.com/camaraproject/Commonalities/blob/86e5c2ab963ea4054b3ee7e7bdda261e64713bd8/documentation/API-design-guidelines.md#9-architecture-headers) states that X-Correlator header is **not Required in OAS Definition**.
Some clarifications are currently proposed in https://github.com/camaraproject/Commonalities/pull/88

BTW, `X-Correlator` or `x-correlator` or more popular header: `X-Correlation-ID`

File: code/API_definitions/carrier_billing.yaml
Comment (rartych) [inline code/API_definitions/carrier_billing.yaml:860]: Without quotes it is rendered by Swager as integer, just: 1
```suggestion
          example: '1.0'
```

File: code/API_definitions/carrier_billing.yaml
Comment (rartych) [inline code/API_definitions/carrier_billing.yaml:887]: Here we should give one example or multiple examples  - it should be updated in Commonalities

It can be:

-  Absolute URI like: https://notificationSendServer12.supertelco.com starting with **uri-scheme**
-  Relative Reference like: /supertelco/notification/SendServer12
```suggestion
example: https://notificationSendServer12.supertelco.com)
```

File: code/API_definitions/carrier_billing.yaml
Comment (bigludo7) [inline code/API_definitions/carrier_billing.yaml:153]: Should be: `application/cloudevents+json:`

File: code/API_definitions/carrier_billing.yaml
Comment (bigludo7) [inline code/API_definitions/carrier_billing.yaml:330]: Should be `application/cloudevents+json:`

**Review by bigludo7 — COMMENTED** [review]
Look good. A small fix for the application/cloudevents+json that we have in the notification. I will take a second look on the state engine later today.

File: code/API_definitions/carrier_billing.yaml
Comment (PedroDiez) [reply code/API_definitions/carrier_billing.yaml:860]: Thanks Rafał, i will set double quote "" (to be aligned with other examples in API Spec)

File: code/API_definitions/carrier_billing.yaml
Comment (PedroDiez) [reply code/API_definitions/carrier_billing.yaml:887]: Fine to have one example. Will also add "" as being a string

File: code/API_definitions/carrier_billing.yaml
Comment (PedroDiez) [reply code/API_definitions/carrier_billing.yaml:153]: Rationale about this value is this? @bigludo7 : https://github.com/camaraproject/Commonalities/blob/2420cf352633eb1076ee5383963fec2e4387605c/artifacts/notification-as-cloud-event.yaml#L79

**Review by bigludo7 — APPROVED** [review]
Look fine for me . Thanks !

File: code/API_definitions/carrier_billing.yaml
Comment (bigludo7) [reply code/API_definitions/carrier_billing.yaml]: Yes and here also: https://github.com/camaraproject/Commonalities/pull/86

**Review by rartych — APPROVED** [review]
LGTM

File: code/API_definitions/carrier_billing.yaml
Comment (PedroDiez) [reply code/API_definitions/carrier_billing.yaml]: Thanks for clarification!

**Pull request merged**
